quixstreams/models/topics/manager.py

```diff
@@ -0,0 +1,446 @@
+import logging
+import pprint
+import warnings
+from typing import Dict, List, Mapping, Optional, Set, Literal
+
+from quixstreams.models.serializers import DeserializerType, SerializerType
+from quixstreams.utils.dicts import dict_values
+from .admin import TopicAdmin
+from .exceptions import (
+    MissingTopicAdmin,
+    MissingTopicForChangelog,
+    TopicValidationError,
+)
+from .topic import Topic, TopicConfig, TimestampExtractor
+
+logger = logging.getLogger(__name__)
+
+__all__ = ("TopicManager",)
+
+
+def affirm_ready_for_create(topics: List[Topic]):
+    """
+    Validate a list of topics is ready for creation attempt
+
+    :param topics: list of `Topic`s
+    """
+    if invalid := [topic.name for topic in topics if not topic.config]:
+        raise ValueError(f"configs for Topics {invalid} were NoneTypes")
+
+
+class TopicManager:
+    """
+    The source of all topic management with quixstreams.
+
+    Generally initialized and managed automatically by an `Application`,
+    but allows a user to work with it directly when needed, such as using it alongside
+    a plain `Producer` to create its topics.
+
+    See methods for details.
+    """
+
+    _topic_partitions = 2
+    _topic_replication = 1
+
+    _topic_extra_config_defaults = {}
+    _changelog_extra_config_defaults = {"cleanup.policy": "compact"}
+    _changelog_extra_config_imports_defaults = {"retention.bytes", "retention.ms"}
+
+    def __init__(
+        self,
+        topic_admin: Optional[TopicAdmin] = None,
+        create_timeout: int = 60,
+    ):
+        """
+        :param topic_admin: an `Admin` instance (required for some functionality)
+        :param create_timeout: timeout for topic creation
+        """
+        self._admin = topic_admin
+        self._topics: Dict[str, Topic] = {}
+        self._changelog_topics: Dict[str, Dict[str, Topic]] = {}
+        self._create_timeout = create_timeout
+
+    @property
+    def topics(self) -> Dict[str, Topic]:
+        return self._topics
+
+    @property
+    def topics_list(self) -> List[Topic]:
+        return dict_values(self._topics)
+
+    @property
+    def changelog_topics(self) -> Dict[str, Dict[str, Topic]]:
+        """
+        Note: `Topic`s are the changelogs.
+
+        returns: the changelog topic dict, {topic_name: {suffix: Topic}}
+        """
+        return self._changelog_topics
+
+    @property
+    def changelog_topics_list(self) -> List[Topic]:
+        return dict_values(self._changelog_topics)
+
+    @property
+    def all_topics(self) -> List[Topic]:
+        return self.topics_list + self.changelog_topics_list
+
+    @property
+    def pretty_formatted_topic_configs(self) -> str:
+        """
+        Returns a print-friendly version of all the topics and their configs
+
+        :return: a pprint-formatted string of all the topics
+        """
+        return pprint.pformat(
+            {
+                topic.name: topic.config.__dict__ if topic.config else "NO CONFIG"
```

Comment:
Similar to other places, please use `dataclasses.asdict()` method to convert a dataclass to a dictionary.